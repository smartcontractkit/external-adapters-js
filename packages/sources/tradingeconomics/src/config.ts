import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as config } from '@chainlink/types'

export type Config = config & {
  client: {
    key: string
    secret: string
  }
}

export const NAME = 'TRADINGECONOMICS'

export const DEFAULT_API_ENDPOINT = 'https://api.tradingeconomics.com/markets'
export const DEFAULT_WS_API_ENDPOINT = 'ws://stream.tradingeconomics.com/'

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix)
  config.api.baseURL = config.api.baseURL || DEFAULT_API_ENDPOINT

  return {
    ...config,
    client: {
      key: util.getRequiredEnv('API_CLIENT_KEY', prefix),
      secret: util.getRequiredEnv('API_CLIENT_SECRET', prefix),
    },
  }
}
