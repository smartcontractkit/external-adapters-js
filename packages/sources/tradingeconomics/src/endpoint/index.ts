import type { TInputParameters as PriceInputParameters } from './price'

export type TInputParameters = PriceInputParameters

export * as price from './price'
export * as price_ws from './price-ws'
