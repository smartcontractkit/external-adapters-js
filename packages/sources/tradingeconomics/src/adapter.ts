import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { AdapterRequest, AdapterResponse, MakeWSHandler } from '@chainlink/types'
import { Config, DEFAULT_WS_API_ENDPOINT, makeConfig, NAME } from './config'

export const customParams = {
  base: ['base', 'from', 'asset'],
}

export const execute = async (input: AdapterRequest, config: Config) => {
  const validator = new Validator(input, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const symbol = (validator.overrideSymbol(NAME) as string).toUpperCase()

  // Fall back to getting the data from HTTP endpoint
  const url = `/symbol/${symbol}`

  const params = {
    c: `${config.client.key}:${config.client.secret}`,
  }

  const request = {
    ...config.api,
    url,
    params,
  }

  const response = await Requester.request(request)
  if (!response.data || response.data.length < 1) {
    throw new Error('no result for query')
  }
  // Replace array by the first object in array
  // to avoid unexpected behavior when returning arrays.
  response.data = response.data[0]

  response.data.result = Requester.validateResultNumber(response.data, ['Last'])
  return Requester.success(jobRunID, response)
}

export const makeExecute = (config?: Config) => {
  return async (request: AdapterRequest) => execute(request, config || makeConfig())
}

export const makeWSHandler = (config?: Config): MakeWSHandler => {
  // http://api.tradingeconomics.com/documentation/Streaming
  // https://github.com/boxhock/tradingeconomics-nodejs-stream/blob/master/src/index.ts
  const withApiKey = (url: string, key: string, secret: string) => `${url}?client=${key}:${secret}`
  const getSubscription = (to: string) => ({ topic: 'subscribe', to })

  return () => {
    const defaultConfig = config || makeConfig()

    return {
      connection: {
        url: withApiKey(
          defaultConfig.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
          defaultConfig.client.key || '',
          defaultConfig.client.secret || '',
        ),
      },
      subscribe: (input) => {
        const validator = new Validator(input, customParams)
        if (validator.error) {
          return
        }
        const base = (validator.overrideSymbol(NAME) as string).toUpperCase()
        return getSubscription(base)
      },
      unsubscribe: () => undefined,
      subsFromMessage: (message) => getSubscription(message?.s),
      isError: (message: any) => Number(message.TYPE) > 400 && Number(message.TYPE) < 900,
      filter: (message) => {
        return message.topic && message.topic !== 'keepalive'
      },
      toResponse: (wsResponse: any): AdapterResponse =>
        Requester.success(undefined, { data: { result: wsResponse?.price } }),
    }
  }
}
