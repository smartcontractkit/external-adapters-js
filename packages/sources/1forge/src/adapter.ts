import { Builder, Logger, Requester, Validator } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
  MakeWSHandler,
} from '@chainlink/ea-bootstrap'
import { DEFAULT_WS_API_ENDPOINT, makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (config?: DefaultConfig): MakeWSHandler => {
  const getSubscription = (symbol?: string, subscribe = true): string => {
    if (!symbol) return ''
    return subscribe ? `subscribe_to|${symbol}` : `unsubscribe_from|${symbol}`
  }
  const getSymbol = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.quotes.inputParameters,
      {},
      { shouldThrowError: false },
    )
    if (validator.error) return
    if (Array.isArray(validator.validated.data.quote)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.quote} supplied as quote. Only non-array tickers can be used for WS`,
      )
      return
    }
    if (Array.isArray(validator.validated.data.base)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.base} supplied as base. Only non-array tickers can be used for WS`,
      )
      return
    }
    const symbol = validator.validated.data.base.toUpperCase()
    const convert = validator.validated.data.quote.toUpperCase()
    return `${symbol}/${convert}`
  }
  const parseResponse = (response: string) => {
    const [type, data] = response.split('|')
    return { type, data: JSON.parse(data) }
  }

  return () => {
    const defaultConfig = config || makeConfig()

    return {
      connection: {
        url: DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => getSubscription(getSymbol(input)),
      unsubscribe: (input) => getSubscription(getSymbol(input), false),
      subsFromMessage: (message: any) => {
        if (!message.data || message.data.indexOf('update') === -1) return
        const { data } = parseResponse(message.data)
        return getSubscription(data.s)
      },
      isError: () => false,
      filter: (message: any) => message.data.indexOf('update') !== -1,
      toResponse: (message: any) => {
        const { data } = parseResponse(message.data)
        return Requester.success('1', { data: { result: data.p } })
      },
      onConnectChain: [
        {
          payload: `login|${defaultConfig.apiKey}`,
          filter: (message: any) => message.data.indexOf('post_login_success') !== -1,
        },
      ],
    }
  }
}
