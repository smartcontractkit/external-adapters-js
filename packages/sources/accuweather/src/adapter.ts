import { Builder } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
