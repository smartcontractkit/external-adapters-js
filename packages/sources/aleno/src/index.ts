import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
