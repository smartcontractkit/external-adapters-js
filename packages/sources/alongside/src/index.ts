import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
