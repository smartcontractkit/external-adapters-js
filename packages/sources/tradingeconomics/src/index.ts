import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute, makeWSHandler, endpointSelector } from './adapter'
import { NAME, makeConfig } from './config'
import * as types from './endpoint'
import rateLimit from './config/limits.json'

const adapterContext = { name: NAME, rateLimit }

const { server } = expose(adapterContext, makeExecute(), makeWSHandler(), endpointSelector)
export { NAME, makeExecute, makeConfig, server, types }
