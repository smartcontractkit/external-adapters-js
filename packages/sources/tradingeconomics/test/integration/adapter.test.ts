import { Requester } from '@chainlink/ea-bootstrap'
import { assertError, assertSuccess } from '@chainlink/ea-test-helpers'
import { AdapterRequest } from '@chainlink/types'
import { makeExecute } from '../../src/adapter'

describe('execute', () => {
  const jobID = '1'
  const execute = makeExecute()
  if (!process.env.API_CLIENT_KEY) process.env.API_CLIENT_KEY = 'test_client_key'
  if (!process.env.API_CLIENT_SECRET) process.env.API_CLIENT_SECRET = 'test_client_secret'
  if (!process.env.SYMBOLS) process.env.SYMBOLS = 'test_symbols'

  describe('successful calls @integration', () => {
    const requests = [
      { name: 'id not supplied', testData: { data: { base: 'FTSE' } } },
      { name: 'base', testData: { id: jobID, data: { base: 'FTSE' } } },
      { name: 'from/to', testData: { id: jobID, data: { from: 'FTSE' } } },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        const data = await execute(req.testData as AdapterRequest)
        assertSuccess({ expected: 200, actual: data.statusCode }, data, jobID)
        expect(data.result).toBeGreaterThan(0)
        expect(data.data.result).toBeGreaterThan(0)
      })
    })
  })

  describe('validation error', () => {
    const requests = [
      { name: 'empty body', testData: {} },
      { name: 'empty data', testData: { data: {} } },
      { name: 'base not supplied', testData: { id: jobID, data: {} } },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        try {
          await execute(req.testData as AdapterRequest)
        } catch (error) {
          const errorResp = Requester.errored(jobID, error)
          assertError({ expected: 400, actual: errorResp.statusCode }, errorResp, jobID)
        }
      })
    })
  })

  describe('error calls @integration', () => {
    const requests = [
      {
        name: 'unknown base',
        testData: { id: jobID, data: { base: 'not_real' } },
      },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        try {
          await execute(req.testData as AdapterRequest)
        } catch (error) {
          const errorResp = Requester.errored(jobID, error)
          assertError({ expected: 500, actual: errorResp.statusCode }, errorResp, jobID)
        }
      })
    })
  })
})
