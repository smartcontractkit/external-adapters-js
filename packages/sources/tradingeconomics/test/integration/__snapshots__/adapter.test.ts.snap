// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute symbol api should return success 1`] = `
{
  "data": {
    "result": 1.15591,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 1.15591,
  "statusCode": 200,
}
`;
